src/sensors/can/motor.hpp

```diff
@@ -0,0 +1,118 @@
+class Motor {
+public:
+    /// @brief Deleted default constructor. Must explicitly construct this object
+    Motor() = delete;
+
+    /// @brief Main constructor. Defines the motor and controller type, global ID, id, and can bus
+    /// @param controller_type The motor controller type 
+    /// @param gid The global ID, not the per-bus motor ID
+    /// @param id The per-bus motor ID. This is 1-indexed 
+    /// @param bus The CAN bus index/ID
+    Motor(MotorControllerType controller_type, uint32_t gid, uint32_t id, uint8_t bus)
+        : m_controller_type(controller_type), m_gid(gid), m_id(id), m_bus_id(bus) {}
+
+    /// @brief Virtual destructor
+    virtual ~Motor() { }
+
+public:
+    /// @brief Initialize the motor and perform any startup commands
+    virtual void init() = 0;
+
+    /// @brief Common read command. Fills given message if successful
+    /// @param msg The message buffer to fill data into
+    /// @return 0 on failure, 1 on success
+    virtual int read(CAN_message_t& msg) = 0;
+
+    /// @brief Common write command. This fills the given message if successful. This is used to compile data to be sent over the CAN line
+    /// @param msg The message buffer to fill data into
+    /// @return The index into msg buf where the motor data was written
+    /// @note Does not issue a CAN command over the bus
+    virtual int write(CAN_message_t& msg) const = 0;
+
+    /// @brief Zero the motor. This is a safety function to ensure the motor is not actively driving
+    /// @note Does not issue a CAN command over the bus
+    virtual void zero_motor() = 0;  // this is virtual because it might differ between motors
+
+    /// @brief Generic write motor function handling only torque. This is the only common interface of all the motors we use
+    /// @param torque The torque value between [-1, 1]
+    /// @note This does not issue a CAN command over the bus
+    virtual void write_motor_torque(float torque) = 0;
+
+    /// @brief Print the current state of the motor
+    virtual void print_state() const = 0;
+
+public:
+    /// @brief Get the motor controller type
+    /// @return The motor controller type
+    inline MotorControllerType get_controller_type() const { return m_controller_type; }
+
+    /// @brief Get the global ID
+    /// @return The unique global ID
+    inline uint32_t get_global_id() const { return m_gid; }
+
+    /// @brief Get the per-bus motor ID
+    /// @return The CAN motor ID
+    inline uint32_t get_id() const { return m_id; }
+
+    /// @brief Get the bus ID
+    /// @return The bus ID
+    inline uint32_t get_bus_id() const { return m_bus_id; }
+
+    /// @brief Get the current state of the motor
+    /// @return The current state of the motor
+    inline MotorState get_state() const { return m_state; }
+
+protected:
+    /// @brief What controller this motor uses
+    MotorControllerType m_controller_type = NULL_MOTOR_CONTROLLER_TYPE;
```

Comment:
discussed that this information is here in case we want to downcast to a motor with more specialized behavior